### PR TITLE
Refact DPA Resource (use new BaseResource)

### PR DESCRIPTION
**Pull request by amastbau**

### CODE REVIEW

File: openshift_resources/oadp/data_protection_application.py [hunk 1]
```diff
@@ -1,17 +1,516 @@
...
+        None,
+        description='CACert defines a CA bundle to use when verifying TLS connections to the provider.',
+    )
+    prefix: Optional[str] = Field(
+        None,
+        description='Prefix is the path inside a bucket to use for Velero storage. Optional.',
+    )
+
+
+class BackupLocationVelero(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:140]: please rename BackupLocationVelero --> VeleroConfig?

File: openshift_resources/oadp/data_protection_application.py [hunk 2]
```diff
@@ -1,17 +1,516 @@
...
+
+class BackupLocation(BaseModel):
+    bucket: Optional[Bucket] = None
+    velero: Optional[BackupLocationVelero] = Field(
+        None,
+        description='BackupStorageLocationSpec defines the desired state of a Velero BackupStorageLocation',
+    )
+
+
+class PodConfigToleration(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:181]: Toleration is from core.v1 package, should be removed from here and import from there instead:
https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/api/core/v1.py#L1584

File: openshift_resources/oadp/data_protection_application.py [hunk 3]
```diff
@@ -1,17 +1,516 @@
...
+        None,
+        description='resourceAllocations defines the CPU and Memory resource allocations for the restic Pod',
+    )
+    tolerations: Optional[List[PodConfigToleration]] = Field(
+        None,
+        description='tolerations defines the list of tolerations to be applied to Restic daemonset',
+    )
+
+
+class Restic(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:220]: Restic -> ResticConfig

File: openshift_resources/oadp/data_protection_application.py [hunk 4]
```diff
@@ -1,17 +1,516 @@
...
+        None, description='timeout defines the Restic timeout, default value is 1h'
+    )
+
+
+class CustomPlugin(BaseModel):
+    image: str
+    name: str
+
+
+class LogLevel(Enum):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:242]: let's move Enums and constants to the dedicated place for constants

File: openshift_resources/oadp/data_protection_application.py [hunk 5]
```diff
@@ -1,17 +1,516 @@
...
+    trace = 'trace'
+    debug = 'debug'
+    info = 'info'
+    warning = 'warning'
+    error = 'error'
+    fatal = 'fatal'
+    panic = 'panic'
+
+
+class TolerationModel(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:252]: please remove, it's redundant. Tolerations should be imported from core.v1:
https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/api/core/v1.py#L1584

File: openshift_resources/oadp/data_protection_application.py [hunk 6]
```diff
@@ -1,17 +1,516 @@
...
+    )
+
+
+class Features(BaseModel):
+    dataMover: Optional[DataMover] = Field(
+        None, description='Contains data mover specific configurations'
+    )
+
+
+class Option(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:345]: what is that class for?

File: openshift_resources/oadp/data_protection_application.py [hunk 7]
```diff
@@ -1,17 +1,516 @@
...
+    name: Optional[str] = Field(None, description='Required.')
+    value: Optional[str] = None
+
+
+class PodDnsConfig(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:350]: PodDnsConfig is defined in core.v1: https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/api/core/v1.py#L896
, please remove and import it from there instead

File: openshift_resources/oadp/data_protection_application.py [hunk 8]
```diff
@@ -1,17 +1,516 @@
...
+
+
+class SnapshotLocation(BaseModel):
+    velero: SnapshotLocationVelero = Field(
+        ...,
+        description='VolumeSnapshotLocationSpec defines the specification for a Velero VolumeSnapshotLocation.',
+    )
+
+
+class DataProtectionApplicationSpecModel(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:381]: please rename DataProtectionApplicationSpecModel -> DataProtectionApplicationSpec

File: openshift_resources/oadp/data_protection_application.py [hunk 9]
```diff
@@ -1,17 +1,516 @@
...
+    )
+
+
+class DataProtectionApplicationConditionConditionStatus(Enum):
+    True_ = 'True'
+    False_ = 'False'
+    Unknown = 'Unknown'
+
+
+class DataProtectionApplicationCondition(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:425]: please remove, condition is suppose to be declared here: https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/apimachinery/pkg/apis/meta/v1.py#L332
please import it

File: openshift_resources/oadp/data_protection_application.py [hunk 10]
```diff
@@ -1,17 +1,516 @@
...
+        regex=r'^([a-z0-9]([-a-z0-9]*[a-z0-9])?(\.[a-z0-9]([-a-z0-9]*[a-z0-9])?)*/)?(([A-Za-z0-9][-A-Za-z0-9_.]*)?[A-Za-z0-9])$',
+        max_length=316,
+    ) = Field(
+        ...,
+        description='type of condition in CamelCase or in foo.example.com/CamelCase. --- Many .condition.type values are consistent across resources like Available, but because arbitrary conditions can be useful (see .node.status.conditions), the ability to deconflict is important. The regex it matches is (dns1123SubdomainFmt/)?(qualifiedNameFmt)',
+    )
+
+
+class DataProtectionApplicationStatus(BaseModel):
+    conditions: Optional[List[DataProtectionApplicationCondition]] = None
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:457]: please use https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/apimachinery/pkg/apis/meta/v1.py#L332 as the elements' type of the list

File: openshift_resources/oadp/data_protection_application.py [hunk 11]
```diff
@@ -1,17 +1,516 @@
...
+        None,
+        description='A list of DNS resolver options. This will be merged with the base options generated from DNSPolicy. Duplicated entries will be removed. Resolution options given in Options will override those that appear in the base DNSPolicy.',
+    )
+    searches: Optional[List[str]] = Field(
+        None,
+        description='A list of DNS search domains for host-name lookup. This will be appended to the base search paths generated from DNSPolicy. Duplicated search paths will be removed.',
+    )
+
+
+class SnapshotLocationVelero(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:365]: please rename SnapshotLocationVelero -> VolumeSnapshotLocationSpec

Also, maybe we should move this to a new resource volume_snapshot_location.py

File: openshift_resources/oadp/data_protection_application.py [hunk 12]
```diff
@@ -1,17 +1,516 @@
...
+    podConfig: Optional[DataProtectionApplicationConfigurationVeleroPodConfig] = Field(
+        None, description='Pod specific configuration'
+    )
+    restoreResourcesVersionPriority: Optional[str] = Field(
+        None,
+        description='restoreResourceVersionPriority represents a configmap that will be created if defined for use in conjunction with EnableAPIGroupVersions feature flag Defining this field automatically add EnableAPIGroupVersions to the velero server feature flag',
+    )
+
+
+class DataProtectionApplicationConfiguration(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:318]: please rename DataProtectionApplicationConfiguration -> ApplicationConfiguration

File: openshift_resources/oadp/data_protection_application.py [hunk 13]
```diff
@@ -1,17 +1,516 @@
...
+        None,
+        description='resourceAllocations defines the CPU and Memory resource allocations for the restic Pod',
+    )
+    tolerations: Optional[List[TolerationModel]] = Field(
+        None,
+        description='tolerations defines the list of tolerations to be applied to Restic daemonset',
+    )
+
+
+class DataProtectionApplicationConfigurationVelero(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:291]: please rename DataProtectionApplicationConfigurationVelero -> VeleroConfig

File: openshift_resources/oadp/data_protection_application.py [hunk 14]
```diff
@@ -1,17 +1,516 @@
...
+        None,
+        description='TolerationSeconds represents the period of time the toleration (which must be of effect NoExecute, otherwise this field is ignored) tolerates the taint. By default, it is not set, which means tolerate the taint forever (do not evict). Zero and negative values will be treated as 0 (evict immediately) by the system.',
+    )
+    value: Optional[str] = Field(
+        None,
+        description='Value is the taint value the toleration matches to. If the operator is Exists, the value should be empty, otherwise just a regular string.',
+    )
+
+
+class DataProtectionApplicationConfigurationVeleroPodConfig(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:275]: redundant, already declared in line 204

File: openshift_resources/oadp/data_protection_application.py [hunk 15]
```diff
@@ -1,17 +1,516 @@
...
+        None,
+        description='credential contains the credential information intended to be used with this location',
+    )
+    default: Optional[bool] = Field(
+        None,
+        description='default indicates this location is the default backup storage location.',
+    )
+
+
+class AccessMode(Enum):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:107]: [PersistentVolumeAccessMode](https://pkg.go.dev/k8s.io/api/core/v1#PersistentVolumeAccessMode) is defined in core v1. If it missing there, let's create move it to the correct location in https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/api/core/v1.py

File: openshift_resources/oadp/data_protection_application.py [hunk 16]
```diff
@@ -1,17 +1,516 @@
...
+import resources.io.k8s.apimachinery.pkg.apis.meta.v1 as v1
 
+from openshift_resources.resource import BaseResource
 
-class DataProtectionApplication(NamespacedResource):
-    api_group = r.ApiGroups.OADP_API_GROUP.value
 
-    def wait_for_successful_reconciliation(self):
-        self.wait_for_condition(condition=r.Condition.CONDITION_RECONCILED.value, status=r.Status.SUC_RECONCILED.value)
+class CreationSecret(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:18]: redundant, defined here already:
https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/api/core/v1.py#L1342

File: openshift_resources/oadp/data_protection_application.py [hunk 17]
```diff
@@ -1,17 +1,516 @@
...
+
+
+class CloudStorageRef(BaseModel):
+    name: Optional[str] = Field(
+        None,
+        description='Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names TODO: Add other useful fields. apiVersion, kind, uid?',
+    )
+
+
+class Credential(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:71]: redundant, defined already in core.v1: https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/api/core/v1.py#L1342

File: openshift_resources/oadp/data_protection_application.py [hunk 18]
```diff
@@ -1,17 +1,516 @@
...
+        description='Name of the referent. More info: https://kubernetes.io/docs/concepts/overview/working-with-objects/names/#names TODO: Add other useful fields. apiVersion, kind, uid?',
+    )
+    optional: Optional[bool] = Field(
+        None, description='Specify whether the Secret or its key must be defined'
+    )
 
-    def wait_for_failed_reconciliation(self):
-        self.wait_for_condition(condition=r.Condition.CONDITION_RECONCILED.value, status=r.Status.FAILED_TO_RECONCILE.value)
 
+class Provider(Enum):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:32]: let's move enums to the dedicated place for constants

File: openshift_resources/oadp/data_protection_application.py [hunk 19]
```diff
@@ -1,17 +1,516 @@
...
+        None,
+        description='snapshotLocations defines the list of desired configuration to use for VolumeSnapshotLocations',
+    )
+    unsupportedOverrides: Optional[Dict[str, str]] = Field(
+        None,
+        description='unsupportedOverrides can be used to override images used in deployments. Available keys are:   - veleroImageFqin   - awsPluginImageFqin   - openshiftPluginImageFqin   - azurePluginImageFqin   - gcpPluginImageFqin   - csiPluginImageFqin   - dataMoverImageFqin   - resticRestoreImageFqin   - kubevirtPluginImageFqin',
+    )
+
+
+class DataProtectionApplicationConditionConditionStatus(Enum):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:419]: ConditionStatus is suppose to be defined in [core v1](https://github.com/oadp-qe/openshift-resources/blob/main/src/resources/io/k8s/api/core/v1.py), if it's missing there, let's add it there and remove from here

https://pkg.go.dev/k8s.io/api/core/v1#ConditionStatus

File: openshift_resources/oadp/data_protection_application.py [hunk 20]
```diff
@@ -1,17 +1,516 @@
...
+    supplementalGroups: Optional[List[int]] = Field(
+        None,
+        description='supplementalGroups defines the linux groups to be applied to the Restic Pod',
+    )
+    timeout: Optional[str] = Field(
+        None, description='timeout defines the Restic timeout, default value is 1h'
+    )
+
+
+class CustomPlugin(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:237]: Let's move Enums to the dedicated place for constants

**Review by mperetzred — CHANGES_REQUESTED** [review]

File: openshift_resources/oadp/data_protection_application.py (hunk 16 above)
Comment (amastbau) [reply openshift_resources/oadp/data_protection_application.py:18]: Cannot find it 
https://pkg.go.dev/search?q=CreationSecret&m=symbol
maybe you mean another class?
@mperetzred

File: openshift_resources/oadp/data_protection_application.py (hunk 13 above)
Comment (amastbau) [reply openshift_resources/oadp/data_protection_application.py:204]: There is this VeleroConfig
https://github.com/vmware-tanzu/velero/blob/v1.9.2/pkg/client/config.go#L38
which is different.
wdyt?
@mperetzred

**Review by amastbau — COMMENTED** [review]
wait for:
https://github.com/oadp-qe/openshift-adp-python-wrapper/pull/31
and #TODO:

File: openshift_resources/oadp/data_protection_application.py (hunk 14 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:188]: please remove

File: openshift_resources/oadp/data_protection_application.py (hunk 11 above)
Comment (amastbau) [reply openshift_resources/oadp/data_protection_application.py:246]: waiting for PR#31

File: openshift_resources/oadp/data_protection_application.py (hunk 13 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:204]: @amastbau you should refer at the DPA API, not the Velero's:
https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#VeleroConfig

File: openshift_resources/oadp/data_protection_application.py (hunk 12 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:231]: don't see you renamed it, so I unresolved the thread

File: openshift_resources/oadp/data_protection_application.py (hunk 10 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:344]: don't see you renamed it, so I unresolved the thread

**Review by mperetzred — CHANGES_REQUESTED** [review]

File: openshift_resources/oadp/data_protection_application.py (hunk 14 above)
Comment (amastbau) [reply openshift_resources/oadp/data_protection_application.py:176]: where do you see it?

File: openshift_resources/oadp/data_protection_application.py [hunk 21]
```diff
@@ -1,17 +1,405 @@
...
+class DataProtectionApplication(BaseResource):
+    apiVersion: Optional[str] = Field(
+        None,
+        description='APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources',
+    )
+    kind: Optional[str] = Field(
+        None,
+        description='Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds',
+    )
+    metadata: Optional[v1.ObjectMeta] = Field(
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:394]: you can remove metadata, as it's already declared in the parent class

File: openshift_resources/oadp/data_protection_application.py [hunk 22]
```diff
@@ -1,17 +1,405 @@
...
+        description='Standard list metadata. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds',
+    )
+
+
+class DataProtectionApplication(BaseResource):
+    apiVersion: Optional[str] = Field(
+        None,
+        description='APIVersion defines the versioned schema of this representation of an object. Servers should convert recognized schemas to the latest internal value, and may reject unrecognized values. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#resources',
+    )
+    kind: Optional[str] = Field(
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:390]: you can remove kind, as it's already declared in the parent class

File: openshift_resources/oadp/data_protection_application.py [hunk 23]
```diff
@@ -1,17 +1,405 @@
...
+        description='Kind is a string value representing the REST resource this object represents. Servers may infer this from the endpoint the client submits requests to. Cannot be updated. In CamelCase. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds',
+    )
+    metadata: Optional[v1.ListMeta] = Field(
+        None,
+        description='Standard list metadata. More info: https://git.k8s.io/community/contributors/devel/sig-architecture/api-conventions.md#types-kinds',
+    )
+
+
+class DataProtectionApplication(BaseResource):
+    apiVersion: Optional[str] = Field(
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:386]: you can remove apiversion, as it's already declared in the parent class

**Review by mperetzred — CHANGES_REQUESTED** [review]

File: openshift_resources/oadp/data_protection_application.py (hunk 14 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:176]: It's called PodConfig, it's the same, and should be imported from core.v1

File: openshift_resources/oadp/data_protection_application.py (hunk 11 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:246]: it's done: https://github.com/oadp-qe/openshift-adp-python-wrapper/pull/31 please review

File: openshift_resources/oadp/data_protection_application.py (hunk 9 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:300]: should use Confition from meta.v1, please remove

File: openshift_resources/oadp/data_protection_application.py [hunk 24]
```diff
@@ -1,17 +1,391 @@
...
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:372]: missing api_group

**Review by mperetzred — CHANGES_REQUESTED** [review]

File: openshift_resources/oadp/data_protection_application.py [hunk 25]
```diff
@@ -1,17 +1,391 @@
...
+        None,
+        description='CACert defines a CA bundle to use when verifying TLS connections to the provider.',
+    )
+    prefix: Optional[str] = Field(
+        None,
+        description='Prefix is the path inside a bucket to use for Velero storage. Optional.',
+    )
+
+
+class VeleroConfig(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:105]: should be called CloudStorageLocation.
 See:
https://pkg.go.dev/github.com/openshift/oadp-operator/api/v1alpha1#CloudStorageLocation

File: openshift_resources/oadp/data_protection_application.py [hunk 26]
```diff
@@ -1,17 +1,391 @@
...
+        None,
+        description='credential contains the credential information intended to be used with this location',
+    )
+    default: Optional[bool] = Field(
+        None,
+        description='default indicates this location is the default backup storage location.',
+    )
+
+
+class VeleroCredential(BaseModel):
```
Comment (mperetzred) [inline openshift_resources/oadp/data_protection_application.py:77]: please remove, should use SecretKeySelector from core.v1

**Review by mperetzred — CHANGES_REQUESTED** [review]

File: openshift_resources/oadp/data_protection_application.py (hunk 14 above)
Comment (mperetzred) [reply openshift_resources/oadp/data_protection_application.py:176]: you haven't changed it yet, but you keep resolving it :)

**Review by mperetzred — CHANGES_REQUESTED** [review]

File: openshift_resources/oadp/data_protection_application.py (hunk 11 above)
Comment (amastbau) [reply openshift_resources/oadp/data_protection_application.py:214]: approved #31

Comment (amastbau) [reply openshift_resources/oadp/data_protection_application.py:214]: @mperetzred

**Pull request closed without merging**